### PR TITLE
Convert seek offsets to `file_off_t`

### PR DESCRIPTION
**Pull request by pwmarcz**
**EDIT**: We introduce our own type, `file_off_t`, instead of using `loff_t`; see discussion.

## Description of the changes 

This converts all internal `off_t` to `loff_t` (plus a few related changes, see commits). 

I'm planning bigger changes to FS regarding file position and size after that (currently we keep it in far too many places), but I want to have the right type first.

Note that this does not make us fully 32-bit ready: we would (at least) need to implement syscalls like `llseek` and `stat64`, because `lseek` and `stat` use `off_t`.

### Why `loff_t`?

This is apparently what Linux uses internally everywhere, for both file size and offset, and I think they have good reasons:

* Initially I wanted to have an usigned type such as `uint64_t`, but it's much easier to check/assert if size or offset is >= 0, than worry about the conversions other way around. With an usigned value, you could end up storing offsets that are too large to be presented to the user (and could be silently converted to negative if we're not careful).
* We talked about using `int64_t`, but `loff_t` is also guaranteed to be 64-bit, and carries more meaning (you know that the variable is for file size/position specifically).

## How to test this PR? 

Not much to check, because there should be no difference on 64-bit systems.

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)

### CODE REVIEW

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 16 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @pwmarcz)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlJ9S89G_aEBIBjVX:-MfHlJ9S89G_aEBIBjVY:b-d4wrch) (no related file):*
Just curious, but was there any reason to have `POLL_SZ` in the first place? Does Linux or some other project do something similar? Where did this idea come from?

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlpYv3kBvjJSm3R_C:-MfHlpYv3kBvjJSm3R_D:b-amgr6h) (no related file):*
Many Jenkins jobs failed, please check.

---
___
*[LibOS/shim/include/shim_fs.h, line 82 at r1](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHjCf20BeRgkXF_925:-MfHjCf20BeRgkXF_926:b-glwefu) ([raw file](https://github.com/oscarlab/graphene/blob/896f0e4add716e274a74c7fec7c67ef82cdc38f6/LibOS/shim/include/shim_fs.h#L82)):*
> ```objc
>     /* poll a single handle */
>     /* POLL_RD|POLL_WR: return POLL_RD|POLL_WR for readable|writable,
>        POLL_ER for failure, -EAGAIN for unknown. */
> ```

Let's remove this comment about `POLL_RD|POLL_WR`. It looks useless.
___
*[LibOS/shim/include/shim_fs.h, line 710 at r1](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHjXxNCzEiZlnX1Bq2:-MfHjXxNCzEiZlnX1Bq3:b-s06vda) ([raw file](https://github.com/oscarlab/graphene/blob/896f0e4add716e274a74c7fec7c67ef82cdc38f6/LibOS/shim/include/shim_fs.h#L710)):*
> ```objc
>  * \param offset desired offset
>  * \param origin `seek` origin parameter (SEEK_SET, SEEK_CUR, SEEK_END)
>  * \param[out] on success, contains new offset
> ```

`new offset` -> `new file position`
___
*[LibOS/shim/src/fs/shim_namei.c, line 117 at r1](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlbs26WP6u59huXTl:-MfHlbs31VM_kfCKLVg4:b-kba4u0) ([raw file](https://github.com/oscarlab/graphene/blob/e921589b871fe31a536dc19b37106b0cf9008938/LibOS/shim/src/fs/shim_namei.c#L117)):*
> ```C
>          *
>          * TODO: remove `mode()` as a separate operation, and make sure this is always done by
>          * `lookup()`.
> ```

How do you make sure that `perm` and `type` are set in `lookup()`? In this PR, I can only see such change for `tmpfs`

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: 14 of 18 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlJ9S89G_aEBIBjVX:-MfHs2Cc6d-7kq92uYpV:b-hl76sg) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Just curious, but was there any reason to have `POLL_SZ` in the first place? Does Linux or some other project do something similar? Where did this idea come from?
</blockquote></details>

No idea. In Graphene, it was there from the beginning of commit history. Initially I thought it's something like "data currently in buffer" (for pipes etc.), but it doesn't seem to be used that way.

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlpYv3kBvjJSm3R_C:-MfIOMI955YjxNqgyCY-:b6i0jen) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Many Jenkins jobs failed, please check.
</blockquote></details>

That was a journey... See commit message for `FS_POLL_SZ`.

---
___
*[LibOS/shim/include/shim_fs.h, line 82 at r1](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHjCf20BeRgkXF_925:-MfIOYIY6hXGHC6WjbAX:bh5xgmi) ([raw file](https://github.com/oscarlab/graphene/blob/896f0e4add716e274a74c7fec7c67ef82cdc38f6/LibOS/shim/include/shim_fs.h#L82)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Let's remove this comment about `POLL_RD|POLL_WR`. It looks useless.
</blockquote></details>

OK. Polling looks like it's in need of rework anyway (then again, that's just one of many things that need reworking).
___
*[LibOS/shim/include/shim_fs.h, line 710 at r1](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHjXxNCzEiZlnX1Bq2:-MfIOgUkBuR20MWvKKdv:b-896fix) ([raw file](https://github.com/oscarlab/graphene/blob/896f0e4add716e274a74c7fec7c67ef82cdc38f6/LibOS/shim/include/shim_fs.h#L710)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`new offset` -> `new file position`
</blockquote></details>

Done.
___
*[LibOS/shim/src/fs/shim_namei.c, line 117 at r1](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlbs26WP6u59huXTl:-MfIOrpi3tbm6s-NQjBe:bxtvdno) ([raw file](https://github.com/oscarlab/graphene/blob/e921589b871fe31a536dc19b37106b0cf9008938/LibOS/shim/src/fs/shim_namei.c#L117)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

How do you make sure that `perm` and `type` are set in `lookup()`? In this PR, I can only see such change for `tmpfs`
</blockquote></details>

That's because other filesystems already fill in this information:

* chroot in `__query_attr` (which gets called by `creat`, `mkdir` and `lookup`),
* tmpfs in `creat` / `mkdir` / `lookup` (`lookup` fills only root, because other files are always created first using `mkdir` and `lookup`),
* pseudo-fs in `lookup` (`creat` and `mkdir` are not supported).

Yes, the code is pretty awful. I am planning to rewrite tmpfs soon (and this PR is preparation for that), and chroot at some point later.

**Review by dimakuv — APPROVED** [review]
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @pwmarcz)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlpYv3kBvjJSm3R_C:-MfIUKeJ6ne8It715CUH:bir5al2) (no related file):*
> `hstat` returned the right size, but `poll(FS_POLL_SZ)` returned 0

To be honest, I don't see the connection between `poll(FS_POLL_SZ) -> 0` and the fact that the mmap becomes shared between processes. If possible, please expand on the comment text. But I don't block since the bug seems quite subtle.

---

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlpYv3kBvjJSm3R_C:-MfIWfaB1I1Uc1I4UU9k:b26uwtn) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

> `hstat` returned the right size, but `poll(FS_POLL_SZ)` returned 0

To be honest, I don't see the connection between `poll(FS_POLL_SZ) -> 0` and the fact that the mmap becomes shared between processes. If possible, please expand on the comment text. But I don't block since the bug seems quite subtle.
</blockquote></details>

Because

1. the checkpointing code ([shim_vma.c](https://github.com/oscarlab/graphene/blob/b8c1afbac34cf5163becc7596869c5708529754a/LibOS/shim/src/bookkeep/shim_vma.c#L1241)) doesn't copy the whole mapping, just the part that belongs to a file (if it determines the file to be shorter than the mapping),
2. the restore code ([shim_vma.c](https://github.com/oscarlab/graphene/blob/b8c1afbac34cf5163becc7596869c5708529754a/LibOS/shim/src/bookkeep/shim_vma.c#L1300)) performs `mmap` on the remaining part of the mapping (starting from `need_mapped` pointer),
3. the LTP test framework ([tst_test.c](https://github.com/linux-test-project/ltp/blob/da2f34028f046a208aa2fed5e287df2538e69f91/lib/tst_test.c#L101)) creates a file, unlinks it, and maps it. Graphene holds an FD to this file and even clones it to child.

On `master`, `get_file_size()` for this file returns 0, so the code behaves as if the file was empty: no data is copied, and the restore code (in child process) calls `mmap` for the whole 4096 bytes.

After my change, `get_file_size()` returns 4096, so 4096 bytes of memory are copied to child process. The child process does not call `mmap` at all.

(I'm not sure if it's worth explaining in commit message... this behavior of mmap/checkpointing seems wrong anyway, but I don't feel qualified to fix it right now).

---

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 14 of 16 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @pwmarcz)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVBHAOCwRptwV3itPK:-MfVBHAOCwRptwV3itPL:bxp33gu) (no related file):*
We should also update `coding-style.rst`, right now it says: "All file offsets and sizes should be stored using `uint64_t`.".

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfVEYNz0QjdpH35msiU:b-gtheb2) (no related file):*
> Why `loff_t`?
> 
> This is apparently what Linux uses internally everywhere, for both file size and offset, and I think they have good reasons:
> 
>   * Initially I wanted to have an usigned type such as `uint64_t`, but it's much easier to check/assert if size or offset is >= 0, than worry about the conversions other way around. With an usigned value, you could end up storing offsets that are too large to be presented to the user (and could be silently converted to negative if we're not careful).
>   * We talked about using `int64_t`, but `loff_t` is also guaranteed to be 64-bit, and carries more meaning (you know that the variable is for file size/position specifically).

One trouble with `loff_t` is that it's just some typedef from stdlib, while we should be in nostdlib "mode". I think this will be easy to solve when fixing the reliance on stdlib, but then I'm not sure if we should stick to the `loff_t` name, this "l" looks a bit weird in context of Graphene (where we don't have a name clash with a legacy `off_t` type). I.e. maybe we should use some custom name to represent the file offset type emulated inside Graphene?

Non-blocking and no strong opinion on this, just want to discuss this a bit and hear others' opinions.

CC: @pwmarcz, @boryspoplawski, @dimakuv.

---
___
*[LibOS/shim/include/shim_fs.h, line 708 at r2](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVAc1CEPXF9F-2sOLi:-MfVAc1DDzLbiwyyb7RX:bs1qkfx) ([raw file](https://github.com/oscarlab/graphene/blob/9477b3f1998805812f2535cb6d4cd6755c4ecbf3/LibOS/shim/include/shim_fs.h#L708)):*
> ```objc
>  * \param offset desired offset
>  * \param origin `seek` origin parameter (SEEK_SET, SEEK_CUR, SEEK_END)
>  * \param[out] on success, contains new file position
> ```

missing param name

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @pwmarcz)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfVL1UtDmPvscqFRKF1:b-m337ou) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

> Why `loff_t`?
> 
> This is apparently what Linux uses internally everywhere, for both file size and offset, and I think they have good reasons:
> 
>   * Initially I wanted to have an usigned type such as `uint64_t`, but it's much easier to check/assert if size or offset is >= 0, than worry about the conversions other way around. With an usigned value, you could end up storing offsets that are too large to be presented to the user (and could be silently converted to negative if we're not careful).
>   * We talked about using `int64_t`, but `loff_t` is also guaranteed to be 64-bit, and carries more meaning (you know that the variable is for file size/position specifically).

One trouble with `loff_t` is that it's just some typedef from stdlib, while we should be in nostdlib "mode". I think this will be easy to solve when fixing the reliance on stdlib, but then I'm not sure if we should stick to the `loff_t` name, this "l" looks a bit weird in context of Graphene (where we don't have a name clash with a legacy `off_t` type). I.e. maybe we should use some custom name to represent the file offset type emulated inside Graphene?

Non-blocking and no strong opinion on this, just want to discuss this a bit and hear others' opinions.

CC: @pwmarcz, @boryspoplawski, @dimakuv.
</blockquote></details>

Since we don't want to rely on libc types and this is an internal type, I would go with something more meaningful. `off64_t` or `file_off_t` come to my mind. Using just `off_t` would be possible (as a typedef int64_t), but it would confuse people, so I wouldn't go that path.
TLDR: I agree, `loff_t` is weird and used for legacy reasons on Linux and I wouldn't copy such naming to Graphene internal types.

---

**Review by dimakuv — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @pwmarcz)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfHlpYv3kBvjJSm3R_C:-MfWS-HMBhgQkTjOzXk6:bvwtkzk) (no related file):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

Because

1. the checkpointing code ([shim_vma.c](https://github.com/oscarlab/graphene/blob/b8c1afbac34cf5163becc7596869c5708529754a/LibOS/shim/src/bookkeep/shim_vma.c#L1241)) doesn't copy the whole mapping, just the part that belongs to a file (if it determines the file to be shorter than the mapping),
2. the restore code ([shim_vma.c](https://github.com/oscarlab/graphene/blob/b8c1afbac34cf5163becc7596869c5708529754a/LibOS/shim/src/bookkeep/shim_vma.c#L1300)) performs `mmap` on the remaining part of the mapping (starting from `need_mapped` pointer),
3. the LTP test framework ([tst_test.c](https://github.com/linux-test-project/ltp/blob/da2f34028f046a208aa2fed5e287df2538e69f91/lib/tst_test.c#L101)) creates a file, unlinks it, and maps it. Graphene holds an FD to this file and even clones it to child.

On `master`, `get_file_size()` for this file returns 0, so the code behaves as if the file was empty: no data is copied, and the restore code (in child process) calls `mmap` for the whole 4096 bytes.

After my change, `get_file_size()` returns 4096, so 4096 bytes of memory are copied to child process. The child process does not call `mmap` at all.

(I'm not sure if it's worth explaining in commit message... this behavior of mmap/checkpointing seems wrong anyway, but I don't feel qualified to fix it right now).
</blockquote></details>

Thanks, understood (the first link immediately shows the problem). No need to explain in the commit message.

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfWSzuN2xlZNAL72gvZ:b-zdh257) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Since we don't want to rely on libc types and this is an internal type, I would go with something more meaningful. `off64_t` or `file_off_t` come to my mind. Using just `off_t` would be possible (as a typedef int64_t), but it would confuse people, so I wouldn't go that path.
TLDR: I agree, `loff_t` is weird and used for legacy reasons on Linux and I wouldn't copy such naming to Graphene internal types.
</blockquote></details>

+1 to `file_off_t`.Though I have no strong opinion on this, but seems like everyone is in favor of having something more descriptive than `loff_t`.

---

**Review by pwmarcz — COMMENTED** [review]
Jenkins, retest this, please (one of the builds has many LTP timeouts, not sure if they're related)

> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574#-:-MfJ9mlJ7bWUf_W42QkL:babi8k3)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski and @pwmarcz)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfX7EDzD89ULfefXUFW:bycf15h) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

+1 to `file_off_t`.Though I have no strong opinion on this, but seems like everyone is in favor of having something more descriptive than `loff_t`.
</blockquote></details>

I like `file_off_t` too. `off64_t` is already used by stdlib for `lseek64` in 32-bit mode, and suggests historical baggage similar to `loff_t`. And `file_off_t` is self-explanatory and distant enough from all the alternatives. :)

I'll change the code and rewrite the commit message.

---

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: 7 of 19 files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski, @dimakuv, and @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfXIPY93fG-juGhwtk4:b-896fix) (no related file):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

I like `file_off_t` too. `off64_t` is already used by stdlib for `lseek64` in 32-bit mode, and suggests historical baggage similar to `loff_t`. And `file_off_t` is self-explanatory and distant enough from all the alternatives. :)

I'll change the code and rewrite the commit message.
</blockquote></details>

Done.

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVBHAOCwRptwV3itPK:-MfXIfW06yIqA5inpyHC:b-896fix) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

We should also update `coding-style.rst`, right now it says: "All file offsets and sizes should be stored using `uint64_t`.".
</blockquote></details>

Done.

---
___
*[LibOS/shim/include/shim_fs.h, line 708 at r2](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVAc1CEPXF9F-2sOLi:-MfXIQKb7WFAJ4-ByoFP:bt2vr1b) ([raw file](https://github.com/oscarlab/graphene/blob/9477b3f1998805812f2535cb6d4cd6755c4ecbf3/LibOS/shim/include/shim_fs.h#L708)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

missing param name
</blockquote></details>

Fixed.

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: 7 of 20 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv and @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfXZBdW3E3QRKzxVLWS:b-vj41fb) (no related file):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I wasn't against `loff_t` because of it's weird name, not because it was picked for historical reasons. Anyway `file_off_t` sounds good.

---

**Review by dimakuv — DISMISSED** [review]
> Reviewed 13 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required) (waiting on @mkow)

**Review by boryspoplawski — COMMENTED** [review]
> Reviewed 1 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required) (waiting on @mkow)

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required) (waiting on @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfXf2yABi8WCEf84o6Z:b-r5bdv6) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

I wasn't against `loff_t` because of it's weird name, not because it was picked for historical reasons. Anyway `file_off_t` sounds good.
</blockquote></details>

Sorry, typo, I meant: "I was against `loff_t` because of it's weird name, not because it was picked for historical reasons."

---

**Review by mkow — APPROVED** [review]
> Reviewed 13 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: :shipit: complete! all files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfXff8x6YjEaMSNV5RE:bwq0eaz) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Sorry, typo, I meant: "I was against `loff_t` because of it's weird name, not because it was picked for historical reasons."
</blockquote></details>

But this weird name is actually because of the historical reasons :)

---

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: :shipit: complete! all files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfXg6cV55XqSLM2RSip:b8cdv6j) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

But this weird name is actually because of the historical reasons :)
</blockquote></details>

Which is irrelevant in our case.

---

**Review by boryspoplawski — APPROVED** [review]

**Review by mkow — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: :shipit: complete! all files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2574#-MfVEYNz0QjdpH35msiT:-MfXhDQCCo67Lx_vJO9f:bb77ufg) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Which is irrelevant in our case.
</blockquote></details>

Yes.

---

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 13 files at r3.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: all files reviewed, all discussions resolved, not enough approvals from different teams (1 more required, approved so far: ITL)

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2574)* status: :shipit: complete! all files reviewed, all discussions resolved
